Avoid db connections when logging db connection errors

nfebe opened the pull request:
`\OC\Log\LogDetails::logDetails` depends on `\OC_App::getAppVersions()` which makes a database connection causing the logger to break when the database service is unavaiable.

Resolves: #37424

nfebe committed:
Replace ILogger method calls index.php

This commit replaces more ILogger method calls with
 `Psr\Log\LoggerInterface` as we gradually move away from the
 custom ILogger.

This commit aslo, sets the customPsrLogger to not depend on the
 database to `logfile` config value.

Signed-off-by: fenn-cs <fenn25.fn@gmail.com>

Fenn-CS commented:
*As part of this PR, I decided to equally move away from our `ILogger` since it's deprecated*

> The ILogger is marked as deprecated with a decision to now use the more stadard `Psr\Log\LoggerInterface`.
> 
> This commit move all logging dependent on our custom
>  ILogger to `Psr\Log\LoggerInterface`
> 
> See : https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-3-logger-interface.md#3-psrlogloggerinterface

The required changes regarding the above is much bigger project that I expected so this is work in progress but it would nice to have feedback it this is the right path.

cc: @ChristophWurst (Thank you in advance)

ChristophWurst commented:
I do not thing that this fixes https://github.com/nextcloud/server/issues/37424 but it's an important and appreciated code cleanup :+1: 

We have to do this change carefully. There is still a lot of code to rely on ILogger, even if has been deprecated for a while. But we don't want to break apps, yet.

I would suggest to only continue replacing ILogger method calls with LoggerInterface for now, and leave the ILogger constants in place.

ChristophWurst commented:
> I noticed that too, and my intention was to do app refactoring-next and append to this commit.

Note that only a few apps live inside this repos and are tightly bundled with Nextcloud server: https://github.com/nextcloud/server/tree/master/apps. The rest lives in independent repositories.

We will not be able to replace all ILogger usages at once. Let's try to migrate this repo and app repos independently. And once if looks like ILogger isn't use any longer we drop it from the public API found in the `\OCP` namespace.

ChristophWurst commented on lib/public/BackgroundJob/IJob.php:
```diff
@@ -55,7 +55,7 @@ interface IJob {
 	 * with the ILogger interface
 	 */
-	public function execute(IJobList $jobList, ILogger $logger = null);
+	public function execute(IJobList $jobList, LoggerInterface $logger = null);
```
We can't do this backwards incompatible change *yet*. Apps that implement `IJob` would have to adjust when going from Nextcloud 26 to Nextcloud 27. In the same way apps that want to support a wider range of Nextcloud, like 25-27, would not be able to exist.

The interface method has already been marked for removal in Nextcloud 25. We can drop it in Nextcloud 28.

Fenn-CS replied:
Make sense! I would backtrack and stick with your earlier recommendation!

ChristophWurst commented on lib/public/Log/ILogFactory.php:
```diff
@@ -40,15 +39,6 @@ interface ILogFactory {
-	 * @param string $path
-	 * @return ILogger
-	 * @since 14.0.0
-	 * @deprecated use \OCP\Log\ILogFactory::getCustomPsrLogger
```
unrelated to your change but the method lacks info about the version it was deprecated

needs research on whether it is ok to remove the method already. we keep deprecated APIs for at least three major versions.

ChristophWurst reviewed (changes requested):
Public APIs are tricky :)

Fenn-CS commented on lib/private/Server.php:
```diff
@@ -1834,6 +1834,13 @@ public function getLogger() {
 
+	public function getPsrLogger(): LoggerInterface {
+		$logFactory = $this->get(LogFactory::class);
+		$default = $this->getConfig()->getSystemValue('datadirectory', \OC::$SERVERROOT . '/data') . '/nextcloud.log';
```
`getConfig()` and `getSystemConfig()` methods are marked as deprecated, what new method is available to achieve this?

Fenn-CS replied:
Turns out `getSystemValue()` and `getAppValue()` all depend on a working database connection so it causes the issue I reported  in my self review.

Is there a way to read the values from `config.php` that does not require the db set up?

miaulalala replied:
you're probably going to read the `config.php` directly. At this point, the database is already a goner and it's just about handling the error gracefully, so that is unusual but the only way to do it.

nfebe reviewed (commented):
Overall, this catches the DB error and displays something like

```
Fatal error: Uncaught Doctrine\DBAL\Exception: Failed to connect to the database: An exception occurred in the driver: SQLSTATE[08006] [7] could not connect to server: Connection refused Is the server running on host "localhost" (127.0.0.1) and accepting TCP/IP connections on port 5432? in /home/fenn/nextcloud/nextcloud/lib/private/DB/Connection.php:142 Stack trace: #0 /home/fenn/nextcloud/nextcloud/3rdparty/doctrine/dbal/src/Connection.php(1531): OC\DB\Connection->connect() #1 /home/fenn/nextcloud/nextcloud/3rdparty/doctrine/dbal/src/Connection.php(1029): Doctrine\DBAL\Connection->getWrappedConnection() #2 /home/fenn/nextcloud/nextcloud/lib/private/DB/Connection.php(264): Doctrine\DBAL\Connection->executeQuery() #3 /home/fenn/nextcloud/nextcloud/3rdparty/doctrine/dbal/src/Query/QueryBuilder.php(345): OC\DB\Connection->executeQuery() #4 /home/fenn/nextcloud/nextcloud/lib/private/DB/QueryBuilder/QueryBuilder.php(280): Doctrine\DBAL\Query\QueryBuilder->execute() #5 /home/fenn/nextcloud/nextcloud/lib/private/AppConfig.php(418): OC\DB\QueryBuilder\QueryBuilder->execute() #6 /home/fenn/nextcloud/nextcloud/lib/private/AppConfig.php(226): OC\AppConfig->loadConfigValues() #7 /home/fenn/nextcloud/nextcloud/lib/private/AllConfig.php(217): OC\AppConfig->getValue() #8 /home/fenn/nextcloud/nextcloud/lib/private/Server.php(1840): OC\AllConfig->getAppValue() #9 /home/fenn/nextcloud/nextcloud/index.php(77): OC\Server->getPsrLogger() #10 {main} thrown in /home/fenn/nextcloud/nextcloud/lib/private/DB/Connection.php on line 142
```

instead of the regular

```
Internal Server Error

The server encountered an internal error and was unable to complete your request.
Please contact the server administrator if this error reappears multiple times, please include the technical details below in your report.
More details can be found in the server log.
```

I am also able to see the error in `/var/log/apache2/error.log`

My expectation was that the `customPsrLogger` would log to `$dataDir.'nextcloud.log'` but nothing is written on that log.

Finally I didn't expect to logger to interrupt the sending regular error message that points to the logs.

Fenn-CS commented:
So after setting a `Psr` compliant logger in `index.php` it appears to me like the only way to write to the log file is to use one of these methods `OC\Log\Errorlog->write()` or `\OC\Log\File->write()` in the `catch` section of the db connect function where this error is thrown (and caught in `index.php` without being logged)

https://github.com/nextcloud/server/blob/8a79636bcb407229a677f12f4ececaeaefdcb5b4/lib/private/DB/Connection.php#L138-L141


So something like 

```php
		} catch (Exception $e) {
			$logFilePath = \OC::getLogFile();
			$logFile = new File($logFilePath, '', $this->systemConfig);
			$logFile->write('app', 'Log testing', 3); // Throws memory limit exceeded
                        // Additionally
			$connectionErrorLogger = \OC::$server->get(LogFactory::class);
			$connectionErrorLogger = $connectionErrorLogger->getCustomPsrLogger($logFilePath);
			$connectionErrorLogger->error("Test Error Message", []); // Throws memory limit exceeded
			 throw new Exception('Failed to connect to the database: ' . $e->getMessage(), $e->getCode());
		}
```

Essentially calling various log methods in the `OC\DB\Connection->connect()` would lead to a memory limit exceeded error.

```Fatal error: Allowed memory size of 134217728 bytes exhausted (tried to allocate 20480 bytes) in /home/fenn/nextcloud/nextcloud/3rdparty/doctrine/dbal/src/Driver/API/PostgreSQL/ExceptionConverter.php on line 81```

Questions
---
- Is this something that anyone has seen and what could be the issue? 
- Are my conclusions about the methods necessary to write to log file true?

Thank you! 

cc : @miaulalala @ChristophWurst

nfebe committed:
Avoid db connection when logging db connection errors

`\OC\Log\LogDetails::logDetails` depends on `\OC_App::getAppVersions()`
which makes a database connection causing the logger to break when the database
service is unavaiable.

Signed-off-by: fenn-cs <fenn25.fn@gmail.com>

Fenn-CS commented:
Update (now fixed)
--- 
I added the commit :

```
Avoid making db connection when logging about database connection

For various reasons, sometimes a database would not be possible. Our file
 logger before this commit has a function call that relies on a db connection
 that is ironic and creates some kind of chicken and egg scenario. This commit,
resolves that.
```

### How does log file look when db there's a db connection problem?

```
"app":"no app in context","message":{"Exception":"Doctrine\\DBAL\\Exception","Message":"Failed to connect to the database: An exception occurred in the driver: SQLSTATE[08006] [7] could not connect to server: Connection refused\n\tIs the server running on host \"localhost\" (127.0.0.1) and accepting\n\tTCP\/IP connections on port 5432?","Code":7,"Trace":[{"file":"\/home\/fenn\/nextcloud\/nextcloud\/3rdparty\/doctrine\/dbal\/src\/Connection.php","line":1531,"function":"connect","class":"OC\\DB\\Connection","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/3rdparty\/doctrine\/dbal\/src\/Connection.php","line":1029,"function":"getWrappedConnection","class":"Doctrine\\DBAL\\Connection","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/DB\/Connection.php","line":285,"function":"executeQuery","class":"Doctrine\\DBAL\\Connection","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/3rdparty\/doctrine\/dbal\/src\/Query\/QueryBuilder.php","line":345,"function":"executeQuery","class":"OC\\DB\\Connection","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/DB\/QueryBuilder\/QueryBuilder.php","line":280,"function":"execute","class":"Doctrine\\DBAL\\Query\\QueryBuilder","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppConfig.php","line":418,"function":"execute","class":"OC\\DB\\QueryBuilder\\QueryBuilder","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppConfig.php","line":184,"function":"loadConfigValues","class":"OC\\AppConfig","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppConfig.php","line":374,"function":"getApps","class":"OC\\AppConfig","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/legacy\/OC_App.php","line":801,"function":"getValues","class":"OC\\AppConfig","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/Server.php","line":731,"function":"getAppVersions","class":"OC_App","type":"::"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppFramework\/Utility\/SimpleContainer.php","line":171,"function":"OC\\{closure}","class":"OC\\Server","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/3rdparty\/pimple\/pimple\/src\/Pimple\/Container.php","line":122,"function":"OC\\AppFramework\\Utility\\{closure}","class":"OC\\AppFramework\\Utility\\SimpleContainer","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppFramework\/Utility\/SimpleContainer.php","line":138,"function":"offsetGet","class":"Pimple\\Container","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/ServerContainer.php","line":171,"function":"query","class":"OC\\AppFramework\\Utility\\SimpleContainer","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppFramework\/Utility\/SimpleContainer.php","line":65,"function":"query","class":"OC\\ServerContainer","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppFramework\/Utility\/SimpleContainer.php","line":193,"function":"get","class":"OC\\AppFramework\\Utility\\SimpleContainer","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppFramework\/Utility\/SimpleContainer.php","line":171,"function":"OC\\AppFramework\\Utility\\{closure}","class":"OC\\AppFramework\\Utility\\SimpleContainer","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/3rdparty\/pimple\/pimple\/src\/Pimple\/Container.php","line":118,"function":"OC\\AppFramework\\Utility\\{closure}","class":"OC\\AppFramework\\Utility\\SimpleContainer","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppFramework\/Utility\/SimpleContainer.php","line":138,"function":"offsetGet","class":"Pimple\\Container","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/ServerContainer.php","line":171,"function":"query","class":"OC\\AppFramework\\Utility\\SimpleContainer","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppFramework\/Utility\/SimpleContainer.php","line":65,"function":"query","class":"OC\\ServerContainer","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/Server.php","line":1111,"function":"get","class":"OC\\AppFramework\\Utility\\SimpleContainer","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppFramework\/Utility\/SimpleContainer.php","line":171,"function":"OC\\{closure}","class":"OC\\Server","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/3rdparty\/pimple\/pimple\/src\/Pimple\/Container.php","line":122,"function":"OC\\AppFramework\\Utility\\{closure}","class":"OC\\AppFramework\\Utility\\SimpleContainer","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppFramework\/Utility\/SimpleContainer.php","line":138,"function":"offsetGet","class":"Pimple\\Container","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/ServerContainer.php","line":171,"function":"query","class":"OC\\AppFramework\\Utility\\SimpleContainer","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppFramework\/Utility\/SimpleContainer.php","line":65,"function":"query","class":"OC\\ServerContainer","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/Server.php","line":2074,"function":"get","class":"OC\\AppFramework\\Utility\\SimpleContainer","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/Files\/View.php","line":119,"function":"getLockingProvider","class":"OC\\Server","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/Server.php","line":465,"function":"__construct","class":"OC\\Files\\View","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppFramework\/Utility\/SimpleContainer.php","line":171,"function":"OC\\{closure}","class":"OC\\Server","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/3rdparty\/pimple\/pimple\/src\/Pimple\/Container.php","line":122,"function":"OC\\AppFramework\\Utility\\{closure}","class":"OC\\AppFramework\\Utility\\SimpleContainer","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppFramework\/Utility\/SimpleContainer.php","line":138,"function":"offsetGet","class":"Pimple\\Container","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/ServerContainer.php","line":171,"function":"query","class":"OC\\AppFramework\\Utility\\SimpleContainer","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppFramework\/Utility\/SimpleContainer.php","line":65,"function":"query","class":"OC\\ServerContainer","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/Server.php","line":1469,"function":"get","class":"OC\\AppFramework\\Utility\\SimpleContainer","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/base.php","line":624,"function":"boot","class":"OC\\Server","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/base.php","line":1181,"function":"init","class":"OC","type":"::"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/index.php","line":35,"args":["\/home\/fenn\/nextcloud\/nextcloud\/lib\/base.php"],"function":"require_once"}],"File":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/DB\/Connection.php","Line":158,"message":"Failed to connect to the database: An exception occurred in the driver: SQLSTATE[08006] [7] could not connect to server: Connection refused\n\tIs the server running on host \"localhost\" (127.0.0.1) and accepting\n\tTCP\/IP connections on port 5432?","exception":{},"CustomMessage":"Failed to connect to the database: An exception occurred in the driver: SQLSTATE[08006] [7] could not connect to server: Connection refused\n\tIs the server running on host \"localhost\" (127.0.0.1) and accepting\n\tTCP\/IP connections on port 5432?"},"level":1}
{"app":"core","message":{"Exception":"Doctrine\\DBAL\\Exception","Message":"Failed to connect to the database: An exception occurred in the driver: SQLSTATE[08006] [7] could not connect to server: Connection refused\n\tIs the server running on host \"localhost\" (127.0.0.1) and accepting\n\tTCP\/IP connections on port 5432?","Code":7,"Trace":[{"file":"\/home\/fenn\/nextcloud\/nextcloud\/3rdparty\/doctrine\/dbal\/src\/Connection.php","line":1531,"function":"connect","class":"OC\\DB\\Connection","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/3rdparty\/doctrine\/dbal\/src\/Connection.php","line":1029,"function":"getWrappedConnection","class":"Doctrine\\DBAL\\Connection","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/DB\/Connection.php","line":285,"function":"executeQuery","class":"Doctrine\\DBAL\\Connection","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/3rdparty\/doctrine\/dbal\/src\/Query\/QueryBuilder.php","line":345,"function":"executeQuery","class":"OC\\DB\\Connection","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/DB\/QueryBuilder\/QueryBuilder.php","line":280,"function":"execute","class":"Doctrine\\DBAL\\Query\\QueryBuilder","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppConfig.php","line":418,"function":"execute","class":"OC\\DB\\QueryBuilder\\QueryBuilder","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppConfig.php","line":184,"function":"loadConfigValues","class":"OC\\AppConfig","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppConfig.php","line":374,"function":"getApps","class":"OC\\AppConfig","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/legacy\/OC_App.php","line":801,"function":"getValues","class":"OC\\AppConfig","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/Server.php","line":731,"function":"getAppVersions","class":"OC_App","type":"::"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppFramework\/Utility\/SimpleContainer.php","line":171,"function":"OC\\{closure}","class":"OC\\Server","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/3rdparty\/pimple\/pimple\/src\/Pimple\/Container.php","line":122,"function":"OC\\AppFramework\\Utility\\{closure}","class":"OC\\AppFramework\\Utility\\SimpleContainer","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppFramework\/Utility\/SimpleContainer.php","line":138,"function":"offsetGet","class":"Pimple\\Container","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/ServerContainer.php","line":171,"function":"query","class":"OC\\AppFramework\\Utility\\SimpleContainer","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppFramework\/Utility\/SimpleContainer.php","line":65,"function":"query","class":"OC\\ServerContainer","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppFramework\/Utility\/SimpleContainer.php","line":193,"function":"get","class":"OC\\AppFramework\\Utility\\SimpleContainer","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppFramework\/Utility\/SimpleContainer.php","line":171,"function":"OC\\AppFramework\\Utility\\{closure}","class":"OC\\AppFramework\\Utility\\SimpleContainer","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/3rdparty\/pimple\/pimple\/src\/Pimple\/Container.php","line":118,"function":"OC\\AppFramework\\Utility\\{closure}","class":"OC\\AppFramework\\Utility\\SimpleContainer","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppFramework\/Utility\/SimpleContainer.php","line":138,"function":"offsetGet","class":"Pimple\\Container","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/ServerContainer.php","line":171,"function":"query","class":"OC\\AppFramework\\Utility\\SimpleContainer","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppFramework\/Utility\/SimpleContainer.php","line":65,"function":"query","class":"OC\\ServerContainer","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/Server.php","line":1111,"function":"get","class":"OC\\AppFramework\\Utility\\SimpleContainer","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppFramework\/Utility\/SimpleContainer.php","line":171,"function":"OC\\{closure}","class":"OC\\Server","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/3rdparty\/pimple\/pimple\/src\/Pimple\/Container.php","line":122,"function":"OC\\AppFramework\\Utility\\{closure}","class":"OC\\AppFramework\\Utility\\SimpleContainer","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppFramework\/Utility\/SimpleContainer.php","line":138,"function":"offsetGet","class":"Pimple\\Container","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/ServerContainer.php","line":171,"function":"query","class":"OC\\AppFramework\\Utility\\SimpleContainer","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppFramework\/Utility\/SimpleContainer.php","line":65,"function":"query","class":"OC\\ServerContainer","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/Server.php","line":2074,"function":"get","class":"OC\\AppFramework\\Utility\\SimpleContainer","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/Files\/View.php","line":119,"function":"getLockingProvider","class":"OC\\Server","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/Server.php","line":465,"function":"__construct","class":"OC\\Files\\View","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppFramework\/Utility\/SimpleContainer.php","line":171,"function":"OC\\{closure}","class":"OC\\Server","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/3rdparty\/pimple\/pimple\/src\/Pimple\/Container.php","line":122,"function":"OC\\AppFramework\\Utility\\{closure}","class":"OC\\AppFramework\\Utility\\SimpleContainer","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppFramework\/Utility\/SimpleContainer.php","line":138,"function":"offsetGet","class":"Pimple\\Container","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/ServerContainer.php","line":171,"function":"query","class":"OC\\AppFramework\\Utility\\SimpleContainer","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/AppFramework\/Utility\/SimpleContainer.php","line":65,"function":"query","class":"OC\\ServerContainer","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/Server.php","line":1469,"function":"get","class":"OC\\AppFramework\\Utility\\SimpleContainer","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/base.php","line":624,"function":"boot","class":"OC\\Server","type":"->"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/base.php","line":1181,"function":"init","class":"OC","type":"::"},{"file":"\/home\/fenn\/nextcloud\/nextcloud\/index.php","line":35,"args":["\/home\/fenn\/nextcloud\/nextcloud\/lib\/base.php"],"function":"require_once"}],"File":"\/home\/fenn\/nextcloud\/nextcloud\/lib\/private\/DB\/Connection.php","Line":158,"CustomMessage":"--"},"level":1}
```

### About the question in the last update

I am still not sure why calling various methods (I tried with logger methods such  and `Log\File->write()`) would lead to a memory execeeded exception in `DBAL` but my best guess is, those methods use a fair amount of memory and are then called multiples times. Printing a string once from the error handling catch for `OC\DB\Connection.php` leads to the string being printed more than 10 times on output, which implies a db connection is attempted several times for the application is boostraped.

kesselb commented:
> Our file logger before this commit has a function call that relies on a db connection that is ironic and creates some kind of chicken and egg scenario.

Hi, do you know which function is the issue?

artonge commented on lib/base.php:
```diff
@@ -1160,6 +1160,22 @@ protected static function handleAuthHeaders(): void {
+	* self::$config object as it might contain important secrets.
+	*/
+	public static function getLogFilePath(): string {
+		return self::$config->getValue('logfile', self::$SERVERROOT . '/data' . '/nextcloud.log');
```
Admins can change the location of the data directory, so this might not always work:

```php
  'datadirectory' => '/arbitrary_data_path',
```

Fenn-CS replied:
@artonge I wonder if that would have an effect here because when there's a database connection error, the values set by the admin can't be read except looking at the config file directly.

This looks at the config file and if it's not set uses a default.

kesselb replied:
> the values set by the admin can't be read except by looking at the config file directly.

https://github.com/nextcloud/server/blob/master/lib/private/Config.php
https://github.com/nextcloud/server/blob/master/lib/private/SystemConfig.php

Should work, even with invalid database credentials.

Fenn-CS replied:
Right @artonge, thanks for pointing that out. Now resolved.

@kesselb yep, thanks for that, it is actually what I am currently using.

artonge commented on lib/private/Server.php:
```diff
@@ -1834,6 +1834,13 @@ public function getLogger() {
+	public function getPsrLogger(): LoggerInterface {
+		$logFactory = $this->get(LogFactory::class);
+		$logFile = \OC::$server->get(SystemConfig::class)->getValue('logfile', false);
+		$logFile = $logFile ? $logFile : \OC::$SERVERROOT . '/data/nextcloud.log';
```
Same here

artonge commented on index.php:
```diff
@@ -30,12 +30,16 @@
 	OC::handleRequest();
 } catch (\OC\ServiceUnavailableException $ex) {
-	\OC::$server->getLogger()->logException($ex, ['app' => 'index']);
+	\OC::$server->getPsrLogger()->error($ex->getMessage(), [
+		'exception' => $ex,
+	]);
+	\OC::getLogFile()->write('core', $ex->getMessage(), 3);
```
Why do we need to log twice?

Fenn-CS replied:
The first logger does not actually write to file, the second logger writes to file.

Fenn-CS replied:
I am not sure why methods like `error`, `info` and `warning` from our implementation of the `Psr` `LoggerInterface` would not write to file.

But I realized they don't and that writing to file is only currently possible via [`OC/Log/File`](https://github.com/nextcloud/server/blob/master/lib/private/Log/File.php)

@kesselb Do you know if this is true?

Fenn-CS replied:
I figured this out! Thanks to @artonge

Fenn-CS commented:
@kesselb Sorry for the delayed response the function that does this is `logDetailsAsJSON`.

See : https://github.com/nextcloud/server/blob/master/lib/private/Log/LogDetails.php

kesselb commented:
> @kesselb Sorry for the delayed response the function that does this is `logDetailsAsJSON`.
> 
> See : https://github.com/nextcloud/server/blob/master/lib/private/Log/LogDetails.php

Thank you :+1: 

It could be \OC_User::getUser() in logDetails then.

kesselb commented:
> > @kesselb Sorry for the delayed response the function that does this is `logDetailsAsJSON`.
> > See : https://github.com/nextcloud/server/blob/master/lib/private/Log/LogDetails.php
> 
> Thank you +1
> 
> It could be \OC_User::getUser() in logDetails then.

https://github.com/nextcloud/server/blob/b294edad804f40618a96116845615831302d0357/lib/private/Server.php#L733

The above line causes the issue. getAppVersions requires a working database connection to fetch the app versions. The function is called in the memcache factory :(

Fenn-CS commented:
> The above line causes the issue. getAppVersions requires a working database connection to fetch the app versions. The function is called in the memcache factory :(

Thank you @kesselb .... Are you suggesting that a solution to refactor that method is sort? 

For now, the logic is "manually" create the json entry if `logAsJSON` fails with a database connection error...

```php
		try {
			$entry = $this->logDetailsAsJSON($app, $message, $level);
		} catch (\Doctrine\DBAL\Exception $e) {
		    $entry = json_encode(['app' => $app, 'message' => $message, 'level' => 1]);
}
```
Please let me what course of action you are suggesting.

come-nc commented on lib/private/Server.php:
```diff
@@ -1834,6 +1834,14 @@ public function getLogger() {
 		return $this->get(ILogger::class);
 	}
 
+	public function getPsrLogger(): LoggerInterface {
+		$logFactory = $this->get(LogFactory::class);
+		$logFile = \OC::$server->get(SystemConfig::class)->getValue('logfile', false);
+		$dataDirectory = \OC::$server->get(SystemConfig::class)->getValue('datadirectory', '/data');
+		$logFile = $logFile ? $logFile : \OC::$SERVERROOT . $dataDirectory . '/nextcloud.log';
+		return $logFactory->getCustomPsrLogger($logFile);
+	}
+
```
What is that?
You should use `\OCP\Server::get(LoggerInterface::class)` in index.php. Or does it fail in this specific case for some reason?
```suggestion
```

Fenn-CS replied:
Yes it does not work.... I believe the class would not be loaded yet for use in `index.php` since that's where the boostraping starts?

Trial report :

```

Fenn-CS replied:
Also, it's my observation that calling the `LoggerInterace`'s `error()` or similar functions does not result to it writing on file.

come-nc replied:
But that is how all the rest of the software is writing logs?
Also, ILogger::class and LoggerInterface::class are registered at the same time, so since index.php is using getLogger, it should be able to use get(LoggerInterface::class) as well.

Fenn-CS replied:
@come-nc Sure... I guess the issue those depend on a database connection?

come-nc commented:
If the problem really is the call to `\OC_App::getAppVersions();`, wrapping it in a try/catch should be enough?

come-nc commented:
> If the problem really is the call to `\OC_App::getAppVersions();`, wrapping it in a try/catch should be enough?

This kind of things seem to work for me:
```
diff --git a/lib/private/Server.php b/lib/private/Server.php
index bb4e217efa3..c0071114b8f 100644
--- a/lib/private/Server.php
+++ b/lib/private/Server.php
@@ -730,7 +730,12 @@ class Server extends ServerContainer implements IServerContainer {
 
                        if ($config->getSystemValueBool('installed', false) && !(defined('PHPUNIT_RUN') && PHPUNIT_RUN)) {
                                if (!$config->getSystemValueBool('log_query')) {
-                                       $v = \OC_App::getAppVersions();
+                                       try {
+                                               $v = \OC_App::getAppVersions();
+                                       } catch (\Throwable $t) {
+                                               // If this throws, we have to ignore it to be able to log errors in database connection
+                                               $v = [];
+                                       }
                                } else {
                                        // If the log_query is enabled, we can not get the app versions
                                        // as that does a query, which will be logged and the logging
```

Now the problem is that this silently ignore an error. I would feel better if I understood why it’s not possible to log without building `OC\Memcache\Factory`

come-nc commented on index.php:
```diff
@@ -30,12 +30,15 @@
 	OC::handleRequest();
 } catch (\OC\ServiceUnavailableException $ex) {
-	\OC::$server->getLogger()->logException($ex, ['app' => 'index']);
+	\OC::$server->getPsrLogger()->error($ex->getMessage(), [
+		'exception' => $ex,
+	]);
```
```suggestion
	\OC::$server->get(LoggerInterface::class)->error($ex->getMessage(), [
		'exception' => $ex,
	]);
```

come-nc replied:
Or even better
```suggestion
	\OCP::Server::get(LoggerInterface::class)->error($ex->getMessage(), [
		'exception' => $ex,
	]);
```

Fenn-CS commented on index.php:
```diff
@@ -30,12 +30,15 @@
 	OC::handleRequest();
 } catch (\OC\ServiceUnavailableException $ex) {
-	\OC::$server->getLogger()->logException($ex, ['app' => 'index']);
+	\OC::$server->get(LoggerInterface::class)->error($ex->getMessage(), [
+		'exception' => $ex,
+	]);
```
Alright.

Fenn-CS commented:
> Now the problem is that this silently ignore an error. I would feel better if I understood why it’s not possible to log without building OC\Memcache\Factory

I added a try-catch that would only catch db connection  failures, I guess that's a more convenient work around.

artonge reviewed (commented):
So the fix is in `Server.php` and the other changes in `index.php` are only to modernize how we get the logger right?

Fenn-CS commented:
> So the fix is in Server.php and the other changes in index.php are only to modernize how we get the logger right?

Exactly!

kesselb commented:
> Now the problem is that this silently ignore an error. I would feel better if I understood why it’s not possible to log without building OC\Memcache\Factory

Good question :+1: 

https://github.com/nextcloud/server/blob/fc076271c791998272856b640be4b63cfe3263a2/lib/private/Log/LogDetails.php#L54

https://github.com/nextcloud/server/blob/fc076271c791998272856b640be4b63cfe3263a2/lib/private/AppFramework/Http/Request.php#L123

CsrfTokenManager -> SessionStorage -> ISession 

https://github.com/nextcloud/server/blob/b294edad804f40618a96116845615831302d0357/lib/private/Server.php#L1403-L1405

https://github.com/nextcloud/server/blob/b294edad804f40618a96116845615831302d0357/lib/private/Server.php#L669

\OC\User\Session -> OC\User\Manager -> ICacheFactory 

ICacheFactory then uses `\OC_App::getAppVersions();` for the cache prefix :cry: 

The good news is, that the csrf token manager is not an essential dependency for the request object itself. We can extract the functionality from the request object to an own service.

kesselb commented on lib/private/Server.php:
```diff
@@ -730,7 +730,13 @@ public function __construct($webRoot, \OC\Config $config) {
+					} catch (\Doctrine\DBAL\Exception $e) {
+						// Database service probably unavailable
+						// Probably related to https://github.com/nextcloud/server/issues/37424
+						$v = [];
```
$v = [] means we are using the nextcloud version number as cache prefix. I'm uncertain if we need a cache prefix/a memcache instance after all.  Likely the database is not available, what do we want to cache? 

```suggestion
						return $arrayCacheFactory;
```

I don't know if that's better or worse. Technically, we disable permanent caching when using the arrayCacheFactory and use a temporary cache,

Fenn-CS replied:
I would let you decide on that since you have more experience with this code base and I would implement the conclusion.

Thanks @kesselb

kesselb replied:
Fair question. I think we should return the arrayCacheFactory when the database is unavailable (and we run into the exception block).

What would we cache when the database is unavailable? ;) If the database is available again, the cache prefix is different, hence the old data is stale. A fluent cache provide seems reasonable for this scenario.

Fenn-CS replied:
Thanks @kesselb would do this right away to this can be closed ASAP!

I hear you are not well, I wish you fast recovery! :)

come-nc commented on lib/private/Server.php:
```diff
@@ -730,7 +730,13 @@ public function __construct($webRoot, \OC\Config $config) {
 			if ($config->getSystemValueBool('installed', false) && !(defined('PHPUNIT_RUN') && PHPUNIT_RUN)) {
 				if (!$config->getSystemValueBool('log_query')) {
-					$v = \OC_App::getAppVersions();
+					try {
+						$v = \OC_App::getAppVersions();
+					} catch (\Doctrine\DBAL\Exception $e) {
```
I’m lost, in the call we had, I thought the conclusion was that this was the wrong place to put the try/catch, and that it should be… I can’t remember where.
On the call to getUser in the logger, no?

Fenn-CS replied:
@come-nc Yeah you are right, however by the time you left I had realized `getUser` was not the culprit. It makes not database access AFIAK.

The stack trace points to `\OC_App::getAppVersions()` and catching the exception there worked unlike with `getUser`

Fenn-CS commented:
@ChristophWurst please can you approve this for merging?
